In [ ]:
!unzip /content/RCdata.zip

Archive:  /content/RCdata.zip
  inflating: chefmozaccepts.csv      
  inflating: chefmozcuisine.csv      
  inflating: chefmozhours4.csv       
  inflating: chefmozparking.csv      
  inflating: geoplaces2.csv          
  inflating: rating_final.csv        
  inflating: README                  
  inflating: usercuisine.csv         
  inflating: userpayment.csv         
  inflating: userprofile.csv         


Data dibagi menjadi tiga bagian
1. Restaurants

 - chefmozaccepts.csv
 - chefmozcuisine.csv
 - chefmozhours4.csv
 - chefmozparking.csv
 - geoplaces2.csv

2. Customers
 - usercuisine.csv
 - userpayment.csv
 - userprofile.csv

3. User-Item-Rating
 - rating_final.csv

In [ ]:
import pandas as pd
 
accepts = pd.read_csv('/content/chefmozaccepts.csv')
cuisine = pd.read_csv('/content/chefmozcuisine.csv')
hours = pd.read_csv('/content/chefmozhours4.csv')
parking = pd.read_csv('/content/chefmozparking.csv')
geo = pd.read_csv('/content/geoplaces2.csv', encoding = "ISO-8859-1")
usercuisine = pd.read_csv('/content/usercuisine.csv')
payment = pd.read_csv('/content/userpayment.csv')
profile = pd.read_csv('/content/userprofile.csv')
rating = pd.read_csv('/content/rating_final.csv')
 
print('Jumlah data pembayaran yang diterima restoran: ', len(accepts.placeID.unique()))
print('Jumlah data masakan pada restoran: ', len(cuisine.placeID.unique()))
print('Jumlah data waktu buka restoran: ', len(hours.placeID.unique()))
print('Jumlah data lokasi restoran: ', len(geo.placeID.unique()))
print('Jumlah data masakan pengguna: ', len(usercuisine.userID.unique()))
print('Jumlah data profil pengguna: ', len(profile.userID.unique()))
print('Jumlah data penilaian yang diberikan pengguna: ', len(rating.userID.unique()))
print('Jumlah data penilaian restoran: ', len(rating.placeID.unique()))

Jumlah data pembayaran yang diterima restoran:  615
Jumlah data masakan pada restoran:  769
Jumlah data waktu buka restoran:  694
Jumlah data lokasi restoran:  130
Jumlah data masakan pengguna:  138
Jumlah data profil pengguna:  138
Jumlah data penilaian yang diberikan pengguna:  138
Jumlah data penilaian restoran:  130


Deskripsi file:
- accepts : merupakan jenis pembayaran yang diterima pada restoran tertentu.
- cuisine : merupakan jenis masakan yang disajikan pada restoran.
- hours : merupakan jadwal buka dan tutupnya restoran.
- parking : merupakan ketersediaan tempat parkir pada restoran.
- geo : merupakan letak restoran.
- usercuisine : merupakan jenis masakan dari data pengguna.
- payment : merupakan jenis pembayaran yang dipakai pengguna.
- profile : merupakan data profil pengguna.

Yang akan diekplorasi hanya file accept, cuisine, profile, dan rating.

Variabel cuisine dan rating akan digunakan pada model rekomendasi kita. Sedangkan, variabel accept dan profile hanya untuk melihat bagaimana profile dan jenis pembayaran yang digunakan oleh pengguna. 

# Exploartory Data Anaysis

## Variabel Accepts

In [ ]:
accepts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1314 entries, 0 to 1313
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   1314 non-null   int64 
 1   Rpayment  1314 non-null   object
dtypes: int64(1), object(1)
memory usage: 20.7+ KB


Berdasarkan output di atas, kita dapat mengetahui bahwa file accept.csv memiliki 1314 entri. Ingatlah juga, accepts termasuk ke dalam kategori data Restaurant. Oleh karena itu, terdapat dua variabel di sini, yaitu placeID dan Rpayment. PlaceID merupakan ID restoran, sedangkan Rpayment merupakan jenis pembayaran yang digunakan pada restoran.

In [ ]:
print('Banyak data: ', len(accepts.placeID.unique()))
print('Jenis Pembayaran yang diterima: ', accepts.Rpayment.unique())

Banyak data:  615
Jenis Pembayaran yang diterima:  ['cash' 'VISA' 'MasterCard-Eurocard' 'American_Express' 'bank_debit_cards'
 'checks' 'Discover' 'Carte_Blanche' 'Diners_Club' 'Visa'
 'Japan_Credit_Bureau' 'gift_certificates']


Berdasarkan output diatas, terdapat 615 data restoran yang unik dengan 12 jenis pembayaran yang diterima antara lain: 

['cash', 'VISA', 'MasterCard-Eurocard', 'American_Express', 'bank_debit_cards'
 'checks', 'Discover', 'Carte_Blanche', 'Diners_Club', 'Visa', 'Japan_Credit_Bureau', 'gift_certificates']


## Variabel Cuisine

In [ ]:
cuisine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   placeID   916 non-null    int64 
 1   Rcuisine  916 non-null    object
dtypes: int64(1), object(1)
memory usage: 14.4+ KB


In [ ]:
print('Banyak tipe masakan: ', len(cuisine.Rcuisine.unique()))
print('Tipe masakan: ', cuisine.Rcuisine.unique())

Banyak tipe masakan:  59
Tipe masakan:  ['Spanish' 'Italian' 'Latin_American' 'Mexican' 'Fast_Food' 'Burgers'
 'Dessert-Ice_Cream' 'Hot_Dogs' 'Steaks' 'Asian' 'International'
 'Mongolian' 'Vegetarian' 'Brazilian' 'Cafe-Coffee_Shop' 'Cafeteria'
 'Contemporary' 'Deli-Sandwiches' 'Diner' 'Japanese' 'Sushi' 'Seafood'
 'Chinese' 'Bar' 'Bar_Pub_Brewery' 'Pizzeria' 'Mediterranean' 'American'
 'Family' 'Caribbean' 'African' 'Breakfast-Brunch' 'Regional' 'Afghan'
 'Bakery' 'Game' 'Armenian' 'Vietnamese' 'Korean' 'Thai' 'Barbecue'
 'Polish' 'Dutch-Belgian' 'French' 'German' 'Southwestern' 'Persian'
 'Ethiopian' 'Juice' 'Soup' 'Continental-European' 'Greek' 'Southern'
 'Eastern_European' 'California' 'Bagels' 'Turkish' 'Organic-Healthy'
 'Fine_Dining']


Berdasarkan output diatas, terdapat 59 tipe makanan yang terdapat pada dateset antara lain:

['Spanish' 'Italian' 'Latin_American' 'Mexican' 'Fast_Food' 'Burgers'
 'Dessert-Ice_Cream' 'Hot_Dogs' 'Steaks' 'Asian' 'International'
 'Mongolian' 'Vegetarian' 'Brazilian' 'Cafe-Coffee_Shop' 'Cafeteria'
 'Contemporary' 'Deli-Sandwiches' 'Diner' 'Japanese' 'Sushi' 'Seafood'
 'Chinese' 'Bar' 'Bar_Pub_Brewery' 'Pizzeria' 'Mediterranean' 'American'
 'Family' 'Caribbean' 'African' 'Breakfast-Brunch' 'Regional' 'Afghan'
 'Bakery' 'Game' 'Armenian' 'Vietnamese' 'Korean' 'Thai' 'Barbecue'
 'Polish' 'Dutch-Belgian' 'French' 'German' 'Southwestern' 'Persian'
 'Ethiopian' 'Juice' 'Soup' 'Continental-European' 'Greek' 'Southern'
 'Eastern_European' 'California' 'Bagels' 'Turkish' 'Organic-Healthy'
 'Fine_Dining']


## Variabel Profile

In [ ]:
profile.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138 entries, 0 to 137
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   userID            138 non-null    object 
 1   latitude          138 non-null    float64
 2   longitude         138 non-null    float64
 3   smoker            138 non-null    object 
 4   drink_level       138 non-null    object 
 5   dress_preference  138 non-null    object 
 6   ambience          138 non-null    object 
 7   transport         138 non-null    object 
 8   marital_status    138 non-null    object 
 9   hijos             138 non-null    object 
 10  birth_year        138 non-null    int64  
 11  interest          138 non-null    object 
 12  personality       138 non-null    object 
 13  religion          138 non-null    object 
 14  activity          138 non-null    object 
 15  color             138 non-null    object 
 16  weight            138 non-null    int64  
 1

In [ ]:
profile.head()

userID   latitude   longitude smoker     drink_level dress_preference  \
0  U1001  22.139997 -100.978803  false      abstemious         informal   
1  U1002  22.150087 -100.983325  false      abstemious         informal   
2  U1003  22.119847 -100.946527  false  social drinker           formal   
3  U1004  18.867000  -99.183000  false      abstemious         informal   
4  U1005  22.183477 -100.959891  false      abstemious    no preference   

  ambience transport marital_status        hijos  birth_year    interest  \
0   family   on foot         single  independent        1989     variety   
1   family    public         single  independent        1990  technology   
2   family    public         single  independent        1989        none   
3   family    public         single  independent        1940     variety   
4   family    public         single  independent        1992        none   

           personality  religion      activity  color  weight  budget  height  
0    thrifty-protector      none       student  black      69  medium    1.77  
1  hunter-ostentatious  Catholic       student    red      40     low    1.87  
2          hard-worker  Catholic       student   blue      60     low    1.69  
3          hard-worker      none  professional  green      44  medium    1.53  
4    thrifty-protector  Catholic       student  black      65  medium    1.69

Menarik. Kita mendapatkan berbagai data mulai dari marital_status, birth_year, smoker, religion, budget, dll. Jika kita ingin membangun sistem rekomendasi yang lebih kompleks, mungkin kita perlu menyertakan fitur-fitur di atas pada model. Data profile ini berguna jika kita ingin membuat sistem rekomendasi berdasarkan demografi atau kondisi sosial pengguna. Namun, untuk studi kasus kali ini, kita tidak akan menggunakan data profile pada model.

## Variabel Rating

In [ ]:
rating.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

Dari fungsi rating.head(), kita dapat mengetahui bahwa data rating terdiri dari 5 kolom dengan tiga kategori rating. Kolom-kolom tersebut antara lain:

- userID, merupakan identitas pengguna.
- placeID, merupakan identitas restoran.
- Rating, merupakan data rating untuk restoran.
- Food_rating, merupakan data rating untuk makanan atau masakan di restoran tersebut.
- Service_rating, merupakan data layanan restoran tersebut.

In [ ]:
rating.describe()

placeID       rating  food_rating  service_rating
count    1161.000000  1161.000000  1161.000000     1161.000000
mean   134192.041344     1.199828     1.215332        1.090439
std      1100.916275     0.773282     0.792294        0.790844
min    132560.000000     0.000000     0.000000        0.000000
25%    132856.000000     1.000000     1.000000        0.000000
50%    135030.000000     1.000000     1.000000        1.000000
75%    135059.000000     2.000000     2.000000        2.000000
max    135109.000000     2.000000     2.000000        2.000000

Dari output di atas, diketahui bahwa nilai maksimum rating adalah 2 dan nilai minimumnya adalah 0. Artinya, skala rating berkisar antara 0 hingga 2. 

In [ ]:
print('Jumlah userID: ', len(rating.userID.unique()))
print('Jumlah placeID: ', len(rating.placeID.unique()))
print('Jumlah data rating: ', len(rating))

Jumlah userID:  138
Jumlah placeID:  130
Jumlah data rating:  1161


Dari output diatas, diketahui jumlah user yangm memberikan rating yaitu sabanyak 130 orang, jumlah restorannya yaitu 130 restoran, dan data ratingnya sebanyak 1161 data

# Data Preprocessing

## Menggabungkan seluruh kategori restoran

Kita akan menggabungkan seluruh data pada kategori Restaurant. Sehingga, kita akan menggunakan placeID yang unik sebagai acuan dalam penggabungan ini. Implementasikan kode berikut.

In [ ]:
import numpy as np
 
# Menggabungkan seluruh placeID pada kategori Restaurant
resto_all = np.concatenate((
    accepts.placeID.unique(),
    cuisine.placeID.unique(),
    hours.placeID.unique(),
    parking.placeID.unique(),
    geo.placeID.unique()
))

# Mengurutkan data dan menghapus data yang sama
resto_all = np.sort(np.unique(resto_all))

print('Jumlah seluruh data restoran berdasarkan placeID: ', len(resto_all))

Jumlah seluruh data restoran berdasarkan placeID:  938


Perhatikanlah, setelah file pada kategori Restaurant digabung, kita memiliki 938 restoran yang unik.

## Menggabungkan seluruh kategori user

Kita akan menggabungkan seluruh data pada kategori User. Sehingga, kita akan menggunakan userID yang unik sebagai acuan dalam penggabungan ini.
Implementasikan kode berikut.

In [ ]:
# Menggabungkan seluruh userID
user_all = np.concatenate((
    usercuisine.userID.unique(),
    payment.userID.unique(),
    profile.userID.unique()
))
 
# Menghapus data yang sama kemudian mengurutkannya
user_all = np.sort(np.unique(user_all)) 
 
print('Jumlah seluruh user: ', len(user_all))

Jumlah seluruh user:  138


Perhatikanlah, setelah file pada kategori User digabung, kita memiliki 138 user yang unik dari 938 restoran yang memiliki rating.

## Mengetahui Jumlah Rating

In [ ]:
# Menggabungkan file accepts, geo, parking, hours ke dalam dataframe resto_info 
resto_info = pd.concat([accepts, geo, parking, hours])
 
# Menggabungkan dataframe rating dengan resto_info berdasarkan nilai placeID
resto = pd.merge(rating, resto_info , on='placeID', how='left')
resto

userID  placeID  rating  food_rating  service_rating Rpayment   latitude  \
0     U1077   135085       2            2               2     cash        NaN   
1     U1077   135085       2            2               2      NaN  22.150802   
2     U1077   135085       2            2               2      NaN        NaN   
3     U1077   135085       2            2               2      NaN        NaN   
4     U1077   135085       2            2               2      NaN        NaN   
...     ...      ...     ...          ...             ...      ...        ...   
8073  U1068   132660       0            0               0      NaN  23.752943   
8074  U1068   132660       0            0               0      NaN        NaN   
8075  U1068   132660       0            0               0      NaN        NaN   
8076  U1068   132660       0            0               0      NaN        NaN   
8077  U1068   132660       0            0               0      NaN        NaN   

       longitude                                     the_geom_meter  \
0            NaN                                                NaN   
1    -100.982680  0101000020957F00009F823DA6094858C18A2D4D37F9A4...   
2            NaN                                                NaN   
3            NaN                                                NaN   
4            NaN                                                NaN   
...          ...                                                ...   
8073  -99.164679  0101000020957F00003D7905C9DC8157C13FCD1AB7334E...   
8074         NaN                                                NaN   
8075         NaN                                                NaN   
8076         NaN                                                NaN   
8077         NaN                                                NaN   

                                       name  ...     accessibility   price  \
0                                       NaN  ...               NaN     NaN   
1                    Tortas Locas Hipocampo  ...  no_accessibility  medium   
2                                       NaN  ...               NaN     NaN   
3                                       NaN  ...               NaN     NaN   
4                                       NaN  ...               NaN     NaN   
...                                     ...  ...               ...     ...   
8073  carnitas mata calle Emilio Portes Gil  ...        completely     low   
8074                                    NaN  ...               NaN     NaN   
8075                                    NaN  ...               NaN     NaN   
8076                                    NaN  ...               NaN     NaN   
8077                                    NaN  ...               NaN     NaN   

      url Rambience franchise    area other_services parking_lot  \
0     NaN       NaN       NaN     NaN            NaN         NaN   
1       ?  familiar         f  closed           none         NaN   
2     NaN       NaN       NaN     NaN            NaN      public   
3     NaN       NaN       NaN     NaN            NaN         NaN   
4     NaN       NaN       NaN     NaN            NaN         NaN   
...   ...       ...       ...     ...            ...         ...   
8073    ?  familiar         f  closed           none         NaN   
8074  NaN       NaN       NaN     NaN            NaN        none   
8075  NaN       NaN       NaN     NaN            NaN         NaN   
8076  NaN       NaN       NaN     NaN            NaN         NaN   
8077  NaN       NaN       NaN     NaN            NaN         NaN   

             hours                  days  
0              NaN                   NaN  
1              NaN                   NaN  
2              NaN                   NaN  
3     00:00-00:00;  Mon;Tue;Wed;Thu;Fri;  
4     00:00-00:00;                  Sat;  
...            ...                   ...  
8073           NaN                   NaN  
8074           NaN                   NaN  
8075  00:00-23:30;  Mon;Tue;Wed;Thu;Fr

In [ ]:
# Cek missing value dengan fungsi isnull()
resto.isnull().sum()

userID               0
placeID              0
rating               0
food_rating          0
service_rating       0
Rpayment          5781
latitude          6917
longitude         6917
the_geom_meter    6917
name              6917
address           6917
city              6917
state             6917
country           6917
fax               6917
zip               6917
alcohol           6917
smoking_area      6917
dress_code        6917
accessibility     6917
price             6917
url               6917
Rambience         6917
franchise         6917
area              6917
other_services    6917
parking_lot       6917
hours             4619
days              4619
dtype: int64

In [ ]:
# Menghitung jumlah rating, food_rating, dan service kemudian menggabungkannya berdasarkan placeID
resto.groupby('placeID').sum()

rating  food_rating  service_rating    latitude    longitude
placeID                                                              
132560       12           24               6   95.009216  -396.667653
132561       15           20              20   94.907276  -396.506024
132564       25           25              30   94.923698  -396.580739
132572       90           90              84  332.124707 -1514.890677
132583       24           24              30   75.689162  -396.937328
...         ...          ...             ...         ...          ...
135088       36           42              36  113.256068  -595.319338
135104       48           80              48  166.270875  -694.179039
135106       96           96              96  221.497088 -1009.760928
135108       65           65              55  243.498787 -1110.269437
135109       24           30              18   75.687139  -396.941400

[130 rows x 5 columns]

## Menggabungkan Data dengan Fitur Nama Resto

In [ ]:
# Definisikan dataframe rating ke dalam variabel all_resto_rate
all_resto_rate = rating
all_resto_rate

userID  placeID  rating  food_rating  service_rating
0     U1077   135085       2            2               2
1     U1077   135038       2            2               1
2     U1077   132825       2            2               2
3     U1077   135060       1            2               2
4     U1068   135104       1            1               2
...     ...      ...     ...          ...             ...
1156  U1043   132630       1            1               1
1157  U1011   132715       1            1               0
1158  U1068   132733       1            1               0
1159  U1068   132594       1            1               1
1160  U1068   132660       0            0               0

[1161 rows x 5 columns]

In [ ]:
# Menggabungkan all resto_rate dengan dataframe geo berdasarkan placeID
all_resto_name = pd.merge(all_resto_rate, geo[['placeID','name']], on='placeID', how='left')
 
# Print dataframe all_resto_name
all_resto_name

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
1     U1077   135038       2            2               1   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
...     ...      ...     ...          ...             ...   
1156  U1043   132630       1            1               1   
1157  U1011   132715       1            1               0   
1158  U1068   132733       1            1               0   
1159  U1068   132594       1            1               1   
1160  U1068   132660       0            0               0   

                                       name  
0                    Tortas Locas Hipocampo  
1                     Restaurant la Chalita  
2                           puesto de tacos  
3                   Restaurante Marisco Sam  
4                                      vips  
...                                     ...  
1156                             palomo tec  
1157                   tacos de la estacion  
1158                          Little Cesarz  
1159     tacos de barbacoa enfrente del Tec  
1160  carnitas mata calle Emilio Portes Gil  

[1161 rows x 6 columns]

In [ ]:
all_resto_name.isnull().sum()

userID            0
placeID           0
rating            0
food_rating       0
service_rating    0
name              0
dtype: int64

## Menggabungkan Data dengan Fitur masakan Resto

In [ ]:
# Menggabungkan dataframe cuisine dengan all_resto_name dan memasukkannya ke dalam variabel all_resto 
all_resto = pd.merge(all_resto_name, cuisine, on='placeID', how='left')
all_resto

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
1     U1077   135038       2            2               1   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
...     ...      ...     ...          ...             ...   
1326  U1043   132630       1            1               1   
1327  U1011   132715       1            1               0   
1328  U1068   132733       1            1               0   
1329  U1068   132594       1            1               1   
1330  U1068   132660       0            0               0   

                                       name   Rcuisine  
0                    Tortas Locas Hipocampo  Fast_Food  
1                     Restaurant la Chalita        NaN  
2                           puesto de tacos    Mexican  
3                   Restaurante Marisco Sam    Seafood  
4                                      vips    Mexican  
...                                     ...        ...  
1326                             palomo tec    Mexican  
1327                   tacos de la estacion    Mexican  
1328                          Little Cesarz   Pizzeria  
1329     tacos de barbacoa enfrente del Tec    Mexican  
1330  carnitas mata calle Emilio Portes Gil        NaN  

[1331 rows x 7 columns]

Inilah data yang akan kita gunakan untuk membuat sistem rekomendasi. Berikutnya, mari kita menuju tahapan Data Preparation.

# Data Preparation

In [ ]:
all_resto.isnull().sum()

userID              0
placeID             0
rating              0
food_rating         0
service_rating      0
name                0
Rcuisine          288
dtype: int64

In [ ]:
# Membersihkan missing value dengan fungsi dropna()
all_resto_clean = all_resto.dropna()
all_resto_clean

userID  placeID  rating  food_rating  service_rating  \
0     U1077   135085       2            2               2   
2     U1077   132825       2            2               2   
3     U1077   135060       1            2               2   
4     U1068   135104       1            1               2   
5     U1068   132740       0            0               0   
...     ...      ...     ...          ...             ...   
1325  U1043   132732       1            1               1   
1326  U1043   132630       1            1               1   
1327  U1011   132715       1            1               0   
1328  U1068   132733       1            1               0   
1329  U1068   132594       1            1               1   

                                    name   Rcuisine  
0                 Tortas Locas Hipocampo  Fast_Food  
2                        puesto de tacos    Mexican  
3                Restaurante Marisco Sam    Seafood  
4                                   vips    Mexican  
5          Carreton de Flautas y Migadas    Mexican  
...                                  ...        ...  
1325                  Taqueria EL amigo     Mexican  
1326                          palomo tec    Mexican  
1327                tacos de la estacion    Mexican  
1328                       Little Cesarz   Pizzeria  
1329  tacos de barbacoa enfrente del Tec    Mexican  

[1043 rows x 7 columns]

## Mengurutkan berdasarkan jenis makanan

In [ ]:
# Mengurutkan resto berdasarkan PlaceID kemudian memasukkannya ke dalam variabel fix_resto
fix_resto = all_resto_clean.sort_values('placeID', ascending=True)
fix_resto

userID  placeID  rating  food_rating  service_rating  \
1303  U1087   132560       1            2               1   
1288  U1050   132560       0            2               0   
14    U1067   132560       1            0               0   
42    U1082   132560       0            0               0   
1052  U1013   132572       1            1               0   
...     ...      ...     ...          ...             ...   
438   U1024   135106       1            1               1   
178   U1020   135109       2            2               1   
1071  U1041   135109       1            2               1   
99    U1030   135109       0            0               0   
180   U1051   135109       1            1               1   

                            name   Rcuisine  
1303          puesto de gorditas   Regional  
1288          puesto de gorditas   Regional  
14            puesto de gorditas   Regional  
42            puesto de gorditas   Regional  
1052                Cafe Chaires  Cafeteria  
...                          ...        ...  
438   El Rincón de San Francisco    Mexican  
178                    Paniroles    Italian  
1071                   Paniroles    Italian  
99                     Paniroles    Italian  
180                    Paniroles    Italian  

[1043 rows x 7 columns]

In [ ]:
# Mengecek berapa jumlah fix_resto
len(fix_resto.placeID.unique())

95

In [ ]:
# Mengecek kategori masakan yang unik
fix_resto.Rcuisine.unique()

array(['Regional', 'Cafeteria', 'American', 'Mexican', 'Fast_Food',
       'Italian', 'Armenian', 'Pizzeria', 'Japanese', 'Vietnamese',
       'Family', 'International', 'Game', 'Burgers', 'Bakery', 'Bar',
       'Breakfast-Brunch', 'Bar_Pub_Brewery', 'Mediterranean',
       'Cafe-Coffee_Shop', 'Contemporary', 'Seafood', 'Chinese'],
      dtype=object)

Perhatikanlah, di antara semua kategori masakan pada data, ada satu yang menarik, yaitu kategori masakan bernama ‘Game’. Hal-hal seperti ini kadang kita temukan pada data. Penyebabnya bisa beragam, misalnya, kesalahan input data atau perbedaan istilah di suatu daerah dengan daerah lainnya. Jika menemukan data seperti ini, hal yang perlu kita lakukan adalah mengeksplorasi datanya lebih lanjut dan melakukan analisis. 

Dari proses analisis, kita akan memahami perbedaan penamaan ini terjadi pada data mana saja dan apa yang bisa kita lakukan untuk mengatasinya. Tentu solusi untuk permasalahan seperti ini akan berbeda tergantung kasusnya. Untuk kasus pada proyek ini, kita akan memilih salah satu penamaan kategori saja untuk digunakan pada data.

Nah, Anda tentu penasaran, restoran mana yang memiliki kategori ‘Game’? Mari kita lihat sama-sama.

In [ ]:
# Mengecek kategori masakan Game 
fix_resto[fix_resto['Rcuisine'] == 'Game']

userID  placeID  rating  food_rating  service_rating name Rcuisine
781   U1015   132851       1            1               1  KFC     Game
509   U1052   132851       1            0               2  KFC     Game
708   U1008   132851       1            1               1  KFC     Game
770   U1037   132851       2            2               1  KFC     Game
574   U1069   132851       1            0               0  KFC     Game
1188  U1131   132851       2            2               2  KFC     Game
764   U1111   132851       2            1               0  KFC     Game

In [ ]:
# Mengecek kategori masakan pada nama restoran KFC
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
781   U1015   132851       1            1               1  KFC      Game
508   U1052   132851       1            0               2  KFC  American
780   U1015   132851       1            1               1  KFC  American
509   U1052   132851       1            0               2  KFC      Game
708   U1008   132851       1            1               1  KFC      Game
707   U1008   132851       1            1               1  KFC  American
770   U1037   132851       2            2               1  KFC      Game
769   U1037   132851       2            2               1  KFC  American
1187  U1131   132851       2            2               2  KFC  American
574   U1069   132851       1            0               0  KFC      Game
1188  U1131   132851       2            2               2  KFC      Game
764   U1111   132851       2            1               0  KFC      Game
763   U1111   132851       2            1               0  KFC  American
573   U1069   132851       1            0               0  KFC  American

Ternyata, KFC memiliki dua kategori masakan yang berbeda, ‘Game’ dan ‘American’. Tentu ini harus diperbaiki. Dalam sistem rekomendasi yang akan kita kembangkan, satu restoran memiliki satu kategori masakan. Dalam hal ini, KFC lebih cocok disebut sebagai restoran dengan kategori masakan American. Jadi, kita perlu mengganti kategori ‘Game’ dengan ‘American’. 

In [ ]:
# Mengubah nama kategori masakan ‘Game’ menjadi ‘American’
fix_resto = fix_resto.replace('Game', 'American')
fix_resto[fix_resto['name'] == 'KFC']

userID  placeID  rating  food_rating  service_rating name  Rcuisine
781   U1015   132851       1            1               1  KFC  American
508   U1052   132851       1            0               2  KFC  American
780   U1015   132851       1            1               1  KFC  American
509   U1052   132851       1            0               2  KFC  American
708   U1008   132851       1            1               1  KFC  American
707   U1008   132851       1            1               1  KFC  American
770   U1037   132851       2            2               1  KFC  American
769   U1037   132851       2            2               1  KFC  American
1187  U1131   132851       2            2               2  KFC  American
574   U1069   132851       1            0               0  KFC  American
1188  U1131   132851       2            2               2  KFC  American
764   U1111   132851       2            1               0  KFC  American
763   U1111   132851       2            1               0  KFC  American
573   U1069   132851       1            0               0  KFC  American

Kini, restoran ‘KFC’ telah memiliki satu kategori masakan seperti yang kita harapkan, yaitu ‘American’. Dalam sistem rekomendasi yang kita kembangkan, penting untuk memastikan satu restoran mewakili satu kategori masakan. Tujuannya supaya tidak terjadi dobel atau rangkap kategori dalam satu restoran. Sehingga, sistem dapat merekomendasikan resto berdasarkan kategori masakannya. 

In [ ]:
preparation = fix_resto
preparation

userID  placeID  rating  food_rating  service_rating  \
1303  U1087   132560       1            2               1   
1288  U1050   132560       0            2               0   
14    U1067   132560       1            0               0   
42    U1082   132560       0            0               0   
1052  U1013   132572       1            1               0   
...     ...      ...     ...          ...             ...   
438   U1024   135106       1            1               1   
178   U1020   135109       2            2               1   
1071  U1041   135109       1            2               1   
99    U1030   135109       0            0               0   
180   U1051   135109       1            1               1   

                            name   Rcuisine  
1303          puesto de gorditas   Regional  
1288          puesto de gorditas   Regional  
14            puesto de gorditas   Regional  
42            puesto de gorditas   Regional  
1052                Cafe Chaires  Cafeteria  
...                          ...        ...  
438   El Rincón de San Francisco    Mexican  
178                    Paniroles    Italian  
1071                   Paniroles    Italian  
99                     Paniroles    Italian  
180                    Paniroles    Italian  

[1043 rows x 7 columns]

In [ ]:
# Membuang data duplikat pada variabel preparation
preparation = preparation.drop_duplicates('placeID')
preparation

userID  placeID  rating  food_rating  service_rating  \
1303  U1087   132560       1            2               1   
1052  U1013   132572       1            1               0   
168   U1118   132583       0            0               0   
24    U1107   132584       2            2               2   
1329  U1068   132594       1            1               1   
...     ...      ...     ...          ...             ...   
681   U1095   135086       1            2               1   
175   U1020   135088       1            2               0   
4     U1068   135104       1            1               2   
488   U1004   135106       2            2               2   
178   U1020   135109       2            2               1   

                                    name   Rcuisine  
1303                  puesto de gorditas   Regional  
1052                        Cafe Chaires  Cafeteria  
168                     McDonalds Centro   American  
24                    Gorditas Dona Tota    Mexican  
1329  tacos de barbacoa enfrente del Tec    Mexican  
...                                  ...        ...  
681          Mcdonalds Parque Tangamanga  Fast_Food  
175                    Cafeteria cenidet  Cafeteria  
4                                   vips    Mexican  
488           El Rincón de San Francisco    Mexican  
178                            Paniroles    Italian  

[95 rows x 7 columns]

In [ ]:
# Mengonversi data series ‘placeID’ menjadi dalam bentuk list
resto_id = preparation['placeID'].tolist()
 
# Mengonversi data series ‘Name’ menjadi dalam bentuk list
resto_name = preparation['name'].tolist()
 
# Mengonversi data series ‘Rcuisine’ menjadi dalam bentuk list
resto_cuisine = preparation['Rcuisine'].tolist()
 
print(len(resto_id))
print(len(resto_name))
print(len(resto_cuisine))

95
95
95


In [ ]:
# Membuat dictionary untuk data ‘resto_id’, ‘resto_name’, dan ‘cuisine’
resto_new = pd.DataFrame({
    'id': resto_id,
    'resto_name': resto_name,
    'cuisine': resto_cuisine
})
resto_new

id                          resto_name    cuisine
0   132560                  puesto de gorditas   Regional
1   132572                        Cafe Chaires  Cafeteria
2   132583                    McDonalds Centro   American
3   132584                  Gorditas Dona Tota    Mexican
4   132594  tacos de barbacoa enfrente del Tec    Mexican
..     ...                                 ...        ...
90  135086         Mcdonalds Parque Tangamanga  Fast_Food
91  135088                   Cafeteria cenidet  Cafeteria
92  135104                                vips    Mexican
93  135106          El Rincón de San Francisco    Mexican
94  135109                           Paniroles    Italian

[95 rows x 3 columns]

# Model Development: Content-Based Filtering

In [ ]:
data = resto_new
data.sample(5)

id                            resto_name    cuisine
41  132875                             shi ro ie   Japanese
80  135059  Restaurant Bar Hacienda los Martinez        Bar
29  132845   Cenaduria El Rincón de Tlaquepaque    Mexican
82  135069             Abondance Restaurante Bar        Bar
71  135046               Restaurante El Reyecito  Fast_Food

## TF-IDF Vectorized

Fungsi TF-IDF Vectorizer memiliki beberapa parameter, antara lain:

- min_df: Digunakan untuk menghilangkan term/istilah yang terlalu jarang muncul. Sebagai contoh, jika diset min_df = 0.01, artinya kita mengabaikan term yang muncul kurang dari 1% dalam teks. Jika diset min_df = 5, artinya kita mengabaikan term yang uncul dalam kurang dari 5 dokumen.

- max_df: Digunakan untuk menghilangkan term yang terlalu sering muncul, biasanya merupakan stop words yang telah kita bahas pada materi sebelumnya. Sebagai contoh, jika diset max_df=0.8, artinya kita mengabaikan term yang muncul lebih dari 80% dalam teks. Jika diset 25, artinya kita mengabaikan term yang muncul di lebih dari 25 dokumen.

- sublinear_tf: Berfungsi untuk melakukan scaling dan bernilai boolean dengan default false. Sublinear_tf=True akan mengubah vektor frekuensi menjadi bentuk logaritmik (1+log(tf)). Sehingga dapat menormalisasi bias terhadap teks yang panjang dan teks yang pendek.

- use_idf: Bernilai boolean dengan default True. Parameter ini memungkinkan kita untuk menggunakan Inverse Document Frequency (IDF). Hal ini berarti term yang terlalu sering muncul dalam teks akan diberi skor lebih sedikit dibanding term yang jarang muncul (hanya muncul pada teks yang spesifik saja).

Dalam proses vektorisasi ini, kita menggunakan dua metode yaitu:

- fit_transform: Metode ini mempelajari kosa kata dan Inverse Document Frequency (IDF) yang telah kita bahas sebelumnya. Ia memberikan nilai return berupa document-term matrix.

- transform: Metode ini mentransformasi dokumen ke dalam document-term matrix.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
 
# Inisialisasi TfidfVectorizer
tf = TfidfVectorizer()
 
# Melakukan perhitungan idf pada data cuisine
tf.fit(data['cuisine']) 
 
# Mapping array dari fitur index integer ke fitur nama
tf.get_feature_names() 

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['american',
 'armenian',
 'bar',
 'bar_pub_brewery',
 'breakfast',
 'brunch',
 'burgers',
 'cafe',
 'cafeteria',
 'chinese',
 'coffee_shop',
 'contemporary',
 'family',
 'fast_food',
 'international',
 'italian',
 'japanese',
 'mexican',
 'pizzeria',
 'regional',
 'seafood',
 'vietnamese']

In [ ]:
# Melakukan fit lalu ditransformasikan ke bentuk matrix
tfidf_matrix = tf.fit_transform(data['cuisine']) 
 
# Melihat ukuran matrix tfidf
tfidf_matrix.shape 

(95, 22)

Perhatikanlah, matriks yang kita miliki berukuran (95, 22). Nilai 95 merupakan ukuran data dan 22 merupakan matrik kategori masakan. 

Untuk menghasilkan vektor tf-idf dalam bentuk matriks, kita menggunakan fungsi todense(). Jalankan kode berikut.

In [ ]:
# Mengubah vektor tf-idf dalam bentuk matriks dengan fungsi todense()
tfidf_matrix.todense()

matrix([[0., 0., 0., ..., 1., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [1., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]])

Selanjutnya, mari kita lihat matriks tf-idf untuk beberapa resto (resto_name) dan kategori masakan (cuisine). Terapkan kode berikut.

In [ ]:
# Membuat dataframe untuk melihat tf-idf matrix
# Kolom diisi dengan jenis masakan
# Baris diisi dengan nama resto
 
pd.DataFrame(
    tfidf_matrix.todense(), 
    columns=tf.get_feature_names(),
    index=data.resto_name
).sample(22, axis=1).sample(10, axis=0)

italian  bar_pub_brewery  \
resto_name                                                                     
Pizzeria Julios                                         0.0              0.0   
McDonalds Centro                                        0.0              0.0   
Restaurant and Bar and Clothesline Carlos N Cha...      0.0              0.0   
Tortas y hamburguesas el gordo                          0.0              0.0   
Log Yin                                                 0.0              0.0   
palomo tec                                              0.0              0.0   
Gordas de morales                                       0.0              0.0   
carnitas_mata                                           0.0              0.0   
La Posada del Virrey                                    0.0              0.0   
Cenaduria El Rincón de Tlaquepaque                     0.0              0.0   

                                                    international  family  \
resto_name                                                                  
Pizzeria Julios                                               0.0     0.0   
McDonalds Centro                                              0.0     0.0   
Restaurant and Bar and Clothesline Carlos N Cha...            0.0     0.0   
Tortas y hamburguesas el gordo                                0.0     0.0   
Log Yin                                                       0.0     0.0   
palomo tec                                                    0.0     0.0   
Gordas de morales                                             0.0     0.0   
carnitas_mata                                                 0.0     0.0   
La Posada del Virrey                                          1.0     0.0   
Cenaduria El Rincón de Tlaquepaque                           0.0     0.0   

                                                    regional  brunch  chinese  \
resto_name                                                                      
Pizzeria Julios                                          0.0     0.0      0.0   
McDonalds Centro                                         0.0     0.0      0.0   
Restaurant and Bar and Clothesline Carlos N Cha...       0.0     0.0      0.0   
Tortas y hamburguesas el gordo                           0.0     0.0      0.0   
Log Yin                                                  0.0     0.0      0.0   
palomo tec                                               0.0     0.0      0.0   
Gordas de morales                                        0.0     0.0      0.0   
carnitas_mata                                            0.0     0.0      0.0   
La Posada del Virrey                                     0.0     0.0      0.0   
Cenaduria El Rincón de Tlaquepaque                      0.0     0.0      0.0   

                                                    coffee_shop  pizzeria  \
resto_name                                                                  
Pizzeria Julios                                             0.0       0.0   
McDonalds Centro                                            0.0       0.0   
Restaurant and Bar and Clothesline Carlos N Cha...          0.0       0.0   
Tortas y hamburguesas el gordo                              0.0       0.0   
Log Yin                                                     0.0       0.0   
palomo tec                                                  0.0       0.0   
Gordas de morales                                           0.0       0.0   
carnitas_mata                                               0.0       0.0   
La Posada del Virrey                                        0.0       0.0   
Cenaduria El Rincón de Tlaquepaque                         0.0       0.0   

                                                    vietnamese  ...  cafe  \
resto_name                                                      ...         
Pizzeria Julios                                            0.0  ...   0.0   
McDonalds Centro                     

Output matriks tf-idf di atas menunjukkan Restaurant Restaurant and Bar and Clothesline Carlos N Charlies meupakan restoran dengan kategori bar, sedangkan restoran palomo tec, Gordas de morales, carnitas_mata dan Cenaduria El Rincón de Tlaquepaque	 termasuk restoran dengan kategori mexican. 

Sampai di sini, kita telah berhasil mengidentifikasi representasi fitur penting dari setiap kategori masakan dengan fungsi tfidfvectorizer. Kita juga telah menghasilkan matriks yang menunjukkan korelasi antara jenis masakan dengan restoran. Selanjutnya, kita akan menghitung derajat kesamaan antara satu restoran dengan restoran lainnya untuk menghasilkan kandidat restoran yang akan direkomendasikan.

## Cosine Similarity

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
 
# Menghitung cosine similarity pada matrix tf-idf
cosine_sim = cosine_similarity(tfidf_matrix) 
cosine_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 1., 0.],
       [0., 0., 0., ..., 1., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

Selanjutnya, mari kita lihat matriks kesamaan setiap resto dengan menampilkan nama restoran dalam 5 sampel kolom (axis = 1) dan 10 sampel baris (axis=0). Jalankan kode berikut

In [ ]:
# Membuat dataframe dari variabel cosine_sim dengan baris dan kolom berupa nama resto
cosine_sim_df = pd.DataFrame(cosine_sim, index=data['resto_name'], columns=data['resto_name'])
print('Shape:', cosine_sim_df.shape)
 
# Melihat similarity matrix pada setiap resto
cosine_sim_df.sample(5, axis=1).sample(10, axis=0)

Shape: (95, 95)


resto_name                           Restaurant Familiar El Chino  \
resto_name                                                          
Restaurante de Mariscos la Langosta                           0.0   
crudalia                                                      0.0   
tacos los volcanes                                            0.0   
little pizza Emilio Portes Gil                                0.0   
el pueblito                                                   0.0   
Taqueria EL amigo                                             0.0   
Tortas Locas Hipocampo                                        0.0   
Restaurante y Pescaderia Tampico                              0.0   
McDonalds Centro                                              0.0   
Subway                                                        0.0   

resto_name                           Cabana Huasteca  La Estrella de Dimas  \
resto_name                                                                   
Restaurante de Mariscos la Langosta              0.0                   0.0   
crudalia                                         0.0                   0.0   
tacos los volcanes                               0.0                   0.0   
little pizza Emilio Portes Gil                   0.0                   0.0   
el pueblito                                      1.0                   1.0   
Taqueria EL amigo                                1.0                   1.0   
Tortas Locas Hipocampo                           0.0                   0.0   
Restaurante y Pescaderia Tampico                 0.0                   0.0   
McDonalds Centro                                 0.0                   0.0   
Subway                                           0.0                   0.0   

resto_name                           puesto de gorditas  shi ro ie  
resto_name                                                          
Restaurante de Mariscos la Langosta                 0.0        0.0  
crudalia                                            0.0        0.0  
tacos los volcanes                                  0.0        0.0  
little pizza Emilio Portes Gil                      0.0        0.0  
el pueblito                                         0.0        0.0  
Taqueria EL amigo                                   0.0        0.0  
Tortas Locas Hipocampo                              0.0        0.0  
Restaurante y Pescaderia Tampico                    0.0        0.0  
McDonalds Centro                                    0.0        0.0  
Subway                                              0.0        0.0

Dengan cosine similarity, kita berhasil mengidentifikasi kesamaan antara satu restoran dengan restoran lainnya. Shape (95, 95) merupakan ukuran matriks similarity dari data yang kita miliki. Berdasarkan data yang ada, matriks di atas sebenarnya berukuran 95 restoran x 95 restoran (masing-masing dalam sumbu X dan Y). Artinya, kita mengidentifikasi tingkat kesamaan pada 95 nama restoran. Tapi tentu kita tidak bisa menampilkan semuanya. Oleh karena itu, kita hanya memilih 10 restoran pada baris vertikal dan 5 restoran pada sumbu horizontal seperti pada contoh di atas. 

Angka 1.0 mengindikasikan bahwa restoran pada kolom X (horizontal) memiliki kesamaan dengan restoran pada baris Y (vertikal). Sebagai contoh, restoran el pueblito dan Taqueria EL amigo teridentifikasi sama (similar) dengan restoran Cabana Huasteca	dan La Estrella de Dimas.

## Mendapatkan Rekomendasi

Buat fungsi Rekomendasi Resto berdasarkan kemiripan dataframe.

Parameter:
- nama_resto : tipe data string (str)
      Nama Restoran (index kemiripan dataframe)
- similarity_data : tipe data pd.DataFrame (object)
      Kesamaan dataframe, simetrik, dengan resto sebagai indeks dan kolom
- items : tipe data pd.DataFrame (object)
      Mengandung kedua nama dan fitur lainnya yang digunakan untuk mendefinisikan kemiripan
- k : tipe data integer (int)
      Banyaknya jumlah rekomendasi yang diberikan

Pada index ini, kita mengambil k dengan nilai similarity terbesar pada index matrix yang diberikan (i).

In [ ]:
def resto_recommendations(nama_resto, similarity_data=cosine_sim_df, items=data[['resto_name', 'cuisine']], k=5): 
    # Mengambil data dengan menggunakan argpartition untuk melakukan partisi secara tidak langsung sepanjang sumbu yang diberikan    
    # Dataframe diubah menjadi numpy
    # Range(start, stop, step)
    index = similarity_data.loc[:,nama_resto].to_numpy().argpartition(
        range(-1, -k, -1))
    
    # Mengambil data dengan similarity terbesar dari index yang ada
    closest = similarity_data.columns[index[-1:-(k+2):-1]]
    
    # Drop nama_resto agar nama resto yang dicari tidak muncul dalam daftar rekomendasi
    closest = closest.drop(nama_resto, errors='ignore')
 
    return pd.DataFrame(closest).merge(items).head(k)

Perhatikanlah, dengan menggunakan argpartition, kita mengambil sejumlah nilai k tertinggi dari similarity data (dalam kasus ini: dataframe cosine_sim_df). Kemudian, kita mengambil data dari bobot (tingkat kesamaan) tertinggi ke terendah. Data ini dimasukkan ke dalam variabel closest. Berikutnya, kita perlu menghapus nama_resto yang yang dicari agar tidak muncul dalam daftar rekomendasi. Dalam kasus ini, nanti kita akan mencari resto yang mirip dengan KFC, sehingga kita perlu drop nama_resto KFC agar tidak muncul dalam daftar rekomendasi yang diberikan nanti.  

Selanjutnya, mari kita terapkan kode di atas untuk menemukan rekomendasi restoran yang mirip dengan KFC. Terapkan kode berikut:

In [ ]:
data[data.resto_name.eq('KFC')]

id resto_name   cuisine
31  132851        KFC  American

In [ ]:
# Mendapatkan rekomendasi restoran yang mirip dengan KFC
resto_recommendations('KFC')

resto_name        cuisine
0                VIPS       American
1  tacos los volcanes       American
2     Pizzeria Julios       American
3             Sirlone  International
4    McDonalds Centro       American

Sistem kita memberikan rekomendasi 5 nama restoran dengan kategori ‘cuisine’ American dan satu kategori International.